Broker ingress metrics

grantr opened the pull request:
## Proposed Changes

- Add an OpenCensus metrics server to the Broker ingress process
- Instrument the ingress data path

## Additional details

Currently WIP because the metrics server exists but instrumentation is missing.

I used [the existing knative/pkg metrics support](https://github.com/knative/pkg/tree/master/metrics) as inspiration, but didn't actually import that code because it depends heavily on a configmap in the knative-eventing namespace, and we don't have one of those yet.

/cc @Harwayne @nachocano

nachocano committed:
Adding filtering using k8s label selectors

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new-model

nachocano committed:
Updating trigger example with filters

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new-model

n3wscott committed:
Refactor controller paths (#786)

* Rename controller to reconciler

For consistency with serving.

* Rename eventing to v1alpha1

For consistency with serving.

* Rename reconcile.go to subscription.go

For consistency with Serving.

* Remove unused file

This is a relic of the copy from serving.

* Move reconciler/names.go to its own package

Avoids naming conflict between reconciler structs and reconciler
package.

* Move inmemory channel to pkg/provisioners

* Move provider to same file as reconciler

For consistency with Serving.

* Update package name in contrib channels

Code in reconciler package moved to reconciler/names.

* Update location of inmemory channel controller

Moved to pkg/provisioners.

* Update dependencies and codegen

* Adding a simple test for controller.

* adding another bad test.

* ran hack/update-codegen.sh

* adding time to the channel reconcile timeout.

* Fix links.

nachocano committed:
Updating after review comments. Only doing exact header matching.
In a follow up PR, I plan to include matching of expressions. It turns
our that they are somewhat more involved than expected.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new-model

nachocano committed:
Adding filtering expressions. Currently using LabelSelectors without
validations. Using reflection to set some unexposed fields for now.

vagababov committed:
change the use of map[string]bool to use k8s string sets. (#789)

* change the use of map[string]bool to use k8s string sets.

This is part of https://github.com/knative/serving/issues/3049.
I also cleaned up some interface and linter errors, but this
code definitely needs much more love and care.
But this is outside of the scope.

* back to poor go code

* also kafka

sbezverk committed:
Clean up unit tests to use actual domain name (#797)

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

mattmoor committed:
Manually fixup prettier.io bugs. (#793)

nachocano committed:
Adding corev1.Events to GCP PubSub channel for easier debugging (#795)

* Emitting corev1.Events from gcppubsub channel.

* Fixing UTs and adding verbs to controller

* Updating dispatcher's event names.

nachocano committed:
Merge remote-tracking branch 'upstream/master' into broker-new

nachocano committed:
Changes to compile

nachocano committed:
moving filter

nachocano committed:
Moving cmds to broker

nachocano committed:
updating controller

nachocano committed:
Moving provider and reconciler to reconciler folder, and merging them into one

Harwayne committed:
Merge branch 'master' into broker-new-model

grantr committed:
Switch to new test image names (#800)

PR #725 removed paths from test image names, but the e2e tests were
still generating image URLs using the previous longer paths.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Adding verbs to dispatcher

nachocano committed:
Adding create/update/patch events to gcp-pubsub dispatcher. (#801)

nachocano committed:
Moving back mains to cmd

nachocano committed:
Updating config

nachocano committed:
Moving resources to reconcilers

nachocano committed:
Moving broker folder back where it belongs

Harwayne committed:
Merge pull request #5 from nachocano/broker-new

Refactoring to new folder structure

nbarthwal committed:
Updating test-infra dependency (#802)

nachocano committed:
Merge branch 'broker-new-model' into broker-new

nachocano committed:
Revert "Merge branch 'broker-new-model' into broker-new"

This reverts commit d567412b062233463335c7feea94be795f7f2e15, reversing
changes made to bd6a8fa9c62460ee696854fb44a898e72e249ae2.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into filter-exp

nachocano committed:
Updating headers and headerExpressions to attributes and attributeExpressions

chaodaiG committed:
update deps:test-infra for fixing release.yaml not found error (#805)

continous and presubmit tests are failing due to release/yaml not found error, this is fixed in test-infra repo, update it here

Harwayne committed:
Merge branch 'master' into broker-clean-up

Harwayne committed:
Opt knative-eventing into Istio injection (for those whose Istio is defaulted to opt-out). (#806)

Harwayne committed:
Reconcilers notice when mroe things change.

Harwayne committed:
PR comment.

Harwayne committed:
Remove redundant watch.

Harwayne committed:
Merge pull request #6 from Harwayne/broker-clean-up

Broker clean up

Harwayne committed:
Unit test scaffold.

Harwayne committed:
Merge pull request #7 from Harwayne/broker-unit-test-scaffold

Broker unit test scaffold. They test almost nothing, but they pass, so gives a better starting point.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

Harwayne committed:
Tests for the namespace reconciler.

Harwayne committed:
Merge pull request #8 from Harwayne/ns-test

Tests for the namespace reconciler.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

Harwayne committed:
Broker controller unit tests.

nachocano committed:
Exact matching for filters without using k8s selectors-based syntax.
The matching can only be done on source and type for now.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into filter-exp

nachocano committed:
Removing t3 as we don't have set expressions

mattmoor-sockpuppet committed:
golang format tools (#803)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

nachocano committed:
Updates after code review

sbezverk committed:
Adding header to the message being published (#794)

* Adding header to the message being published

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Adding header to the message being published

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing comment

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing comment

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Adding unit test

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* fixing unit test

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

nachocano committed:
Attempt to reconcile broker in trigger controller

Harwayne committed:
Merge pull request #9 from Harwayne/broker-test

Broker unit tests.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

Commit:
Update README.md (#810)

Harwayne committed:
Merge pull request #4 from nachocano/filter-exp

Trigger filters on event source and type

nachocano committed:
Deleting and re-creating subscription object as the backing channel
spec is immutable.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Adding comment

Commit:
User can specify the namespace in yaml for knative-eventing (#711)

* User can specify the namespace in yaml for knative-eventing

* change ns to knative-testing

* update the test ns in reconcile_test.go

* change the test case for the new namespaces

* add clock.go in system

* Use knatve/pkg

Harwayne committed:
Rename Trigger.Spec.Filter.ExactMatch to Trigger.Spec.Filter.SourceAndType. Add unit tests for the filter binary.

Harwayne committed:
Merge pull request #11 from Harwayne/filter-test

Rename Trigger.Spec.Filter.ExactMatch to Trigger.Spec.Filter.SourceAndType. Add unit tests for the filter binary.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Adding event messages

nachocano committed:
Using the broker's namespacedNamed as key to the triggers map.
With this we allow to reconcile only the triggers that belong to the
particular broker that changed

dprotaso committed:
bump knative/pkg - we drop webhook logic that increments spec.generation (#813)

Harwayne committed:
Merge pull request #10 from nachocano/broker-new

Reconcile Triggers on Broker changes.

mattmoor-sockpuppet committed:
golang format tools (#812)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

nachocano committed:
Adding some tests to trigger

nachocano committed:
More UTs

nachocano committed:
More UTs

nachocano committed:
More UTs

Harwayne committed:
Namespace reconciler automatically creates the Broker Filter's ServiceAccount and RBAC.

Sadly this doesn't work well because we have such an old version of controller-runtime that the Filter ends up trying to watch _all_ Triggers, not just those in its namespace. And it only gets permission for the Triggers in its own namespace.

Harwayne committed:
Merge pull request #12 from nachocano/broker-new-trigger-ut

Trigger controller UTs

Harwayne committed:
Remove no longer needed label.

nachocano committed:
Broker and trigger types UTs

grantr committed:
WIP early E2E test

Harwayne committed:
Switch the default channel provisioner from 'in-memory-channel' to 'in-memory'. (#790)

matzew committed:
:scream: for now, we skip some rules from mdl (#758)

sbezverk committed:
Restart dispatcher on connection loss (#796)

* Restart dispatcher on connection loss

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing comments

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Refactor dispatcher for a dynamic reconnect

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Removing mutes

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Fixing unit test

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Fixing unit tests part 2

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Fixing unit tests part 3 racing issue

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Removing redundant mutex

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Address some issue in the logic

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Fixing logic a bit more

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing comments

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing more comments

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* fix racing in unit test

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Fixing a race

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing comments

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing racing conditions

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing comments

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing more comments

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

* Addressing more comments part3

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

nachocano committed:
Changes after code review. Adding trigger defaults and validation tests

nachocano committed:
Cleaner trigger validation

nachocano committed:
Adding dummy tests for broker validation...
Should be implemented

Harwayne committed:
Merge pull request #14 from nachocano/broker-new-trigger-ut

Broker and trigger types UTs

nachocano committed:
Merge remote-tracking branch 'grantr/broker-new-model-e2e' into broker-e2e

nachocano committed:
Compiling and moving things around

nbarthwal committed:
update dep for github.com/knative/test-infra along with fixes (#820)

nachocano committed:
Updating test

Commit:
Replace DOCKER_REPO_OVERRIDE by KO_DOCKER_REPO (#819)

nachocano committed:
More updates

nachocano committed:
Waiting for potentially multiple contents.
Removing check for corev1.Service ready.
Removing grant's great design. Just making it simpler for now.

sbezverk committed:
Switch to a non blocking send (#821)

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

nachocano committed:
Compiling

nachocano committed:
Fixing compilation

evankanderson committed:
Fix env vars from #711 (#824)

nachocano committed:
Fixing compilation errors. Adding AnnotateNamespace function.

nachocano committed:
Adding ns

nachocano committed:
Adding logs. Changing to lowercase any otherwise the pod name is invalid

nachocano committed:
Removing namespace when creating trigger subscriber spec.
Adding wait time constant for default broker creation.

nachocano committed:
Checking if all triggers are ready

nachocano committed:
Updated logs

nachocano committed:
Working

nachocano committed:
Adding logs... Still not receiving the events.

nachocano committed:
More logs

nachocano committed:
Adding build constraint

nachocano committed:
Removing unnecessary stuff

nachocano committed:
Removing ugly structs

nachocano committed:
More logs

Harwayne committed:
Merge pull request #13 from Harwayne/ns-rbac

Namespace reconciler automatically creates the Broker Filter's ServiceAccount and RBAC.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-e2e

nachocano committed:
Removing quotes

nachocano committed:
More logs

Harwayne committed:
Add Harwayne as an approver. (#782)

nachocano committed:
Adding delay

nachocano committed:
Listing triggers in receiver when we create it, so that we don't miss
any message because the client couldn't find the existing trigger.
This is a problem in the in-memory-channel as it doesn't do retries.
Maybe the right solution is to add that functionality there.

nachocano committed:
Adding delay to sender pod

nachocano committed:
Removing withDelay method and just sleep for a while

nachocano committed:
Improve log...

mattmoor-sockpuppet committed:
golang format tools (#832)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

Harwayne committed:
ConfigMapWatcher.Start() blocks, rather than eagerly returning nil (#828)

* ConfigMapWatcher.Start() blocks, rather than eagerly returning nil.

* Move logging message closer to what it describes.

* Precreate the in-memory configMap.

* Print the time in a human readable format.

* hack/update-deps.sh

* PR comments.

nachocano committed:
Updates after code review.

nachocano committed:
Adding some more logs and trailing dots.

mattmoor-sockpuppet committed:
golang format tools (#838)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

Harwayne committed:
Bump Controller Runtime to 0.1.9 (#833)

* ConfigMapWatcher.Start() blocks, rather than eagerly returning nil.

* Move logging message closer to what it describes.

* Precreate the in-memory configMap.

* Print the time in a human readable format.

* hack/update-deps.sh

* PR comments.

* K8s 1.12.5 and Controller Runtime 0.1.9.

* hack/update-codegen.sh

* Update function call with the newly required parameter.

Harwayne committed:
Add labels to the stateful set. (#829)

Harwayne committed:
Merge branch 'master' into broker-controller-runtime

Harwayne committed:
Switch import order.

Harwayne committed:
Merge pull request #16 from Harwayne/broker-controller-runtime

Merge upstream/master.

Commit:
remove inlined and scothis from reviewers and approvers (#839)

nachocano committed:
Updating comments.

nachocano committed:
Updating comments.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-gen-pass

Harwayne committed:
Replace the bad errgroup usage with the runnableServer.

Harwayne committed:
Merge pull request #18 from Harwayne/codegen

Replace the bad errgroup usage with the runnableServer.

mattmoor-sockpuppet committed:
golang format tools (#822)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

Harwayne committed:
Namespace scoped the Broker Filter's client.

Harwayne committed:
Fix unit tests.

Harwayne committed:
Fix yaml

Harwayne committed:
Merge pull request #19 from Harwayne/namespace-rolebinding

Namespace rolebinding, unit test fixes, and YAML fix.

Harwayne committed:
Remove the fake ListOptions that were used exclusively for limitations in the fake client. #833 updated the fake client to remove the need for these options. (#834)

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-e2e

nachocano committed:
Merge branch 'broker-e2e' of github.com:nachocano/eventing into broker-e2e

nachocano committed:
Setting source to source not type.
Updating comment.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-gen-pass

Harwayne committed:
Merge pull request #15 from nachocano/broker-e2e

Broker-Trigger e2e test

Harwayne committed:
Merge pull request #17 from nachocano/broker-gen-pass

Broker-Trigger general pass

Harwayne committed:
Simplify Channel Watch (#835)

* ConfigMapWatcher.Start() blocks, rather than eagerly returning nil.

* Move logging message closer to what it describes.

* Precreate the in-memory configMap.

* Print the time in a human readable format.

* hack/update-deps.sh

* PR comments.

* K8s 1.12.5 and Controller Runtime 0.1.9.

* hack/update-codegen.sh

* Update function call with the newly required parameter.

* Switch to a simpler Channel Watch.

Harwayne committed:
Switch from annotating the namespace to labeling it, to match Istio.

Harwayne committed:
Merge pull request #20 from Harwayne/ns-eventing-label

Switch from annotating the namespace to labeling it, to match Istio.

adrcunha committed:
Update test-infra (#840)

Fixes [knative/test-infra#513](https://github.com/knative/test-infra/issues/513).

Major changes:
* more features in E2E tests
* properly fail releases when `ko` fails

Harwayne committed:
General clean up.

Harwayne committed:
Merge pull request #21 from Harwayne/broker-cleanup

General clean up.

Harwayne committed:
Merge branch 'master' into broker-merge-master

Harwayne committed:
Merge pull request #22 from Harwayne/broker-merge-master

Merge branch 'master' into broker-new-model

grantr committed:
Move Shutdown to runnableServer

Call http.Server.Shutdown in runnableServer.Start when stopCh is closed.
Allows the lifecycle of runnableServer to be handled entirely by the
Manager, including shutdown.

grantr committed:
Add a prometheus server to broker ingress

grantr committed:
correct method comment

grantr committed:
Add a metrics port to the broker's ingress Service

Ultimately access to this port might need different permissions than the
default (ingress) port.

grantr commented on cmd/broker/ingress/main.go:
```diff
@@ -150,9 +175,29 @@ func (f *Handler) dispatch(msg *provisioners.Message) error {
+func (r *runnableServer) Start(stopCh <-chan struct{}) error {
+	logger := r.logger.With(zap.String("address", r.s.Addr))
+	logger.Info("Listening...")
+	if r.ShutdownTimeout > 0 {
```
@Harwayne I moved the shutdown call here so it's easier to run multiple listeners without juggling multiple shutdowns.

Harwayne replied:
Makes sense to move it, but to honor the Runnable contract (as I understand it), I want this always shutdown the server. Some values may have no timeout on the shutdown, but I do want Shutdown called.

grantr replied:
Now Shutdown is always called, and if ShutdownTimeout is not positive, no timeout is given.

An alternate approach that could make sense is:
- Positive: use as timeout
- Zero: use as timeout (shut down immediately)
- Negative: no timeout

Essentially the choice is: should the default behavior be immediate shutdown or infinite shutdown? I'm not sure which is more reasonable, but I'm leaning toward infinite, because problems with infinite shutdown will be obvious (process never stops, users notice downtime) versus subtle for immediate shutdown (transient failure on restart, users can't reproduce).

grantr committed:
Always shut down runnableServer

If ShutdownTimeout is not positive, call Shutdown anyway without a
timeout.

grantr committed:
Add port values to the ingress container spec

These are informational, and more useful now that two ports are
exposed.

nachocano commented on cmd/broker/ingress/main.go:
```diff
@@ -70,26 +76,45 @@ func main() {
 	}
 
+	// Metrics
+	e, err := prometheus.NewExporter(prometheus.Options{Namespace: componentName})
```
If we have multiple brokers in a namespace or even different brokers in different namespaces, will all be sending metrics as if they were the same component? I'm not aware of exporters, but I guess I got worried about that componentName to be the same for everyone. In other words, would it be easy to distinguish from which broker ingress the metrics came from? Does it make sense?

grantr replied:
Good point, we should probably add a tag for Broker name. I think the easiest way to do this is to add the Broker name to the Pod env vars.

The Prometheus server can also apply tags server-side, but I don't know if namespace is one of those.

nachocano commented on cmd/broker/ingress/main.go:
```diff
@@ -150,9 +175,34 @@ func (f *Handler) dispatch(msg *provisioners.Message) error {
+	// shut down when the stop channel is closed. If this is zero or negative,
+	// shutdown will never time out.
+	// TODO alternative: zero shuts down immediately, negative means infinite
+	ShutdownTimeout time.Duration
```
does this need to be exposed or can be named shutdownTimeout?

Harwayne replied:
`runnableServer` isn't exposed, so I don't think this field needs to be either.

grantr replied:
Exporting this field signals to the user that it's something they should change (and adds the field documentation to godoc). When I see an internal variable like `shutdownTimeout`, I assume it's intended for private use by the object, but that's not the case here.

grantr replied:
Since we'll need this code for all the binaries that export metrics, I plan to move it to a shared place in knative/eventing where it will be exported.

grantr committed:
Instrument message count and dispatch time

Also register the exporter so it can start serving metrics.

grantr committed:
Simplify runnableServer shutdown

No need to select since we're just waiting for the channel to be closed.

Harwayne commented on cmd/broker/ingress/main.go:
```diff
@@ -125,7 +150,15 @@ func NewHandler(logger *zap.Logger, destination string) *Handler {
+		}()
+
 		// TODO Filter.
+		// metricsCtx, _ = tag.New(metricsCtx, tag.Insert(TagResult, "filtered"))
```
Remove.

grantr replied:
Replaced with a description of how to tag the measurement when filtering is added.

Harwayne commented on cmd/broker/ingress/main.go:
```diff
@@ -125,7 +150,15 @@ func NewHandler(logger *zap.Logger, destination string) *Handler {
 func createReceiverFunction(f *Handler) func(provisioners.ChannelReference, *provisioners.Message) error {
 	return func(_ provisioners.ChannelReference, m *provisioners.Message) error {
+		metricsCtx := context.Background()
+		defer func() {
```
Why is this defered?

grantr replied:
This allows tags to be inserted into the metrics context in the body of the method while ensuring that the measurement is always recorded.

Harwayne commented on cmd/broker/ingress/metrics.go:
```diff
@@ -0,0 +1,86 @@
+)
+
+const (
+	metricsNamespace = "broker_ingress"
```
Is it worth including knative?

grantr replied:
Not sure if it's correct, but the prometheus exporter in knative/pkg as used by knative/serving doesn't. The prometheus namespace is just the component name, for example `controller`.

https://github.com/knative/serving/blob/c5d12355921c32c4c140e3e3cc12db6a31868523/cmd/controller/main.go#L59
https://github.com/knative/serving/blob/c5d12355921c32c4c140e3e3cc12db6a31868523/cmd/controller/main.go#L203
https://github.com/knative/pkg/blob/1ed7b993c877a0d1914574344e500f5c99909086/metrics/prometheus_exporter.go#L31

Harwayne commented on cmd/broker/ingress/metrics.go:
```diff
@@ -0,0 +1,86 @@
+	// MeasureMessagesTotal is a counter which records the number of messages received
+	// by the ingress. The value of the Result tag indicates whether the message
+	// was filtered or dispatched.
+	MeasureMessagesTotal = stats.Int64(
```
Can this metric have the broker namespace and name as a parameter? So that we can have the ability to distinguish the number of events that have gone through broker-a and broker-b. 

For ingress in particular, it is tied to a single broker, so we might be able to use some knowledge based on the Pod, but for triggers, we won't have that ability. The code will know which trigger a metric should be recorded against, but not the Pod.

I don't know if it is possible in OpenCensus, tags kind of look like what I am thinking about.

grantr replied:
Based on my reading of the Prometheus docs, namespace is normally added as a tag by its Kubernetes service discovery implementation: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#pod

Serving additionally specifies in its scrape config that the `__meta_kubernetes_namespace` label should be renamed to `namespace`:
https://github.com/knative/serving/blob/d454ec94aa1ec39d9f28539a6a0d6753c67f09ef/config/monitoring/metrics/prometheus/100-prometheus-scrape-config.yaml

Based on this prior art I think it's not necessary to specify the namespace tag, because the Pod namespace will always match. **However** if the broker ingress ever changes to be cluster-wide or work across namespaces, then metrics will need to be manually tagged with the Broker's namespace. In that case the implementation is likely to have access to the Broker object for generating tags.

I'll set the Broker name in an optional environment variable set by the controller.

Harwayne commented on cmd/broker/ingress/main.go:
```diff
@@ -70,26 +75,46 @@ func main() {
 	if err = mgr.Start(stopCh); err != nil {
 		logger.Error("manager.Start() returned an error", zap.Error(err))
 	}
 	logger.Info("Exiting...")
```
I think Go programs exit as soon as main() finishes. As we no longer wait on Shutdown() inside main(), will this forcibly exit, without waiting for shutdown to do so gracefully?

grantr replied:
It will forcibly exit. I refactored this and added a WaitGroup so the process waits for the http servers to shut down. The real fix is to change the controller-runtime Manager to wait for all Runnables to stop before returning from Start, as described in https://github.com/kubernetes-sigs/controller-runtime/issues/350.

grantr committed:
Use a WaitGroup to stop runnableServers

Manager doesn't actually wait for Runnables to stop, so we need to add a
WaitGroup to wait for the HTTP Server Shutdown to complete. This is
hopefully temporary until
https://github.com/kubernetes-sigs/controller-runtime/issues/350 is
fixed.

grantr committed:
Hook up internal controller-runtime logger

grantr committed:
Include GCP auth library

Running outside GCP seems to not work without this.

grantr committed:
Translate commented code into a directive

grantr committed:
Tag measurements with the broker name

The BROKER environment variable may contain the broker name. If
non-empty, measurements will be tagged with the given string.

grantr commented:
RFAL @Harwayne @nachocano

Harwayne commented on cmd/broker/ingress/main.go:
```diff
@@ -64,6 +71,8 @@ func main() {
 
 	c := getRequiredEnv("CHANNEL")
 
+	brokerName, _ = os.LookupEnv("BROKER")
```
Why make this optional? 

Especially because we will only notice it is missing when we start looking at metrics that we hoped had it.

grantr replied:
For convenience in debugging when we don't care about metrics. But I'm fine with making it non-optional, since it does makes the normal case safer.

Harwayne replied:
This is a main(), which we don't traditionally unit test...

grantr replied:
This is required now.

Harwayne commented on cmd/broker/ingress/main.go:
```diff
@@ -115,6 +126,8 @@ func main() {
 		logger.Error("manager.Start() returned an error", zap.Error(err))
 	}
 	logger.Info("Exiting...")
+	wg.Wait()
```
I think this has to be time bounded. I don't want a single odd server to hold up the binary from restarting indefinitely.

grantr replied:
Kubernetes takes care of this for us. The default grace period is 30 seconds, after which a container is forcibly terminated. https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods

If we want native shutdown timeout, something like this might work:

```go
ctx := context.WithTimeout(context.Background(), time.Second*30)
go func() {
  select {
  case <-ctx.Done():
    log.Fatal("shutdown took too long")
  }
}()
```

But I think I prefer relying on Kubernetes in this case, since that gives the operator control over the timeouts without the code being involved.

Harwayne replied:
My preference is to do both. We can set the timeout here to something relatively high, but I like the guarantee of a bound.

In fact, if something in the manager returns an error from its `Start()`, then the manager will initiate the shutdown. This will happen without K8s being aware. So if something else in the manager doesn't shutdown, K8s won't help. Right?

grantr replied:
Good point, I'll add a shutdown timeout here.

My hope is that this will only be necessary until the Manager is fixed to have reasonable shutdown behavior.

Harwayne commented on cmd/broker/ingress/main.go:
```diff
@@ -213,11 +251,12 @@ func (r *runnableServer) Start(stopCh <-chan struct{}) error {
-	}()
-	return r.s.ListenAndServe()
+	}
+	r.wg.Done()
```
I _think_ this might panic if for some reason the stopCh was already closed. As the starting go routine might not have had time to start, so calling Done() may cause it to go below zero, and thus panic.

grantr replied:
Good point. Maybe this should go inside the serving goroutine? I think that would have the same behavior but without the possible panic.

nachocano commented:
/lgtm 
once Adam's comments have been addressed.

grantr commented:
Broker merged, so I'll close this one and create another one on the parent repo.
